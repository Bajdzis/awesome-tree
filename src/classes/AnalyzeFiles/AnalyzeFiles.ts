import { FileContent } from '../FileContent/FileContent';
import { FileContentNode } from '../FileContentNode/FileContentNode';
import { FileContentNodeAnalyze } from '../FileContentNodeAnalyze/FileContentNodeAnalyze';

export class AnalyzeFiles {
    private files: FileContentNodeAnalyze[];
    constructor(){
        this.files = [];
    }

    addFile(file : FileContent) {
        const fileNode = file.getFileGraph();
        const signNodes = FileContentNodeAnalyze.addPathToAllGraph(fileNode, file.getPathInfo());
        this.files.push(signNodes);
    }

    getNumberOfFile(){
        return this.files.length;
    }

    analyze() {
        const file = new FileContentNode();
        return file.setChildren(this.analyzeNodes(this.files));
    }

    private analyzeNodes(nodes: FileContentNodeAnalyze[]): FileContentNodeAnalyze[] {

        const grouped: FileContentNodeAnalyze[][] = [];
        nodes.forEach(node => {
            const indexToAdd = grouped.findIndex(group => group[0].isSame(node));
            if(indexToAdd === -1) {
                grouped.push([node]);
            } else {
                grouped[indexToAdd].push(node);
            }
        });

        return this.attachChildren(grouped);
    }

    private analyzeChildrenNodes(nodes: FileContentNodeAnalyze[][]): FileContentNodeAnalyze[] {

        const grouped: FileContentNodeAnalyze[][] = [];

        nodes.forEach(children => {
            children.forEach(childrenNode => {
                const indexToAdd = grouped.findIndex(group => group[0].isSame(childrenNode) && group.every(node => !node.isSameBasePath(childrenNode)));
                if(indexToAdd === -1) {
                    grouped.push([childrenNode]);
                } else {
                    grouped[indexToAdd].push(childrenNode);
                }
            });
        });

        return this.attachChildren(grouped);
    }

    private attachChildren(grouped: FileContentNodeAnalyze[][]) {
        return grouped.map(nodes => {
            const [node] = nodes;
            const result = this.analyzeChildrenNodes(nodes.map(node => node.getChildren()));
            const newNode = node.setChildren(result);

            return newNode;
        });
    }
}
