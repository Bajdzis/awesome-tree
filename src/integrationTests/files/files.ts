import * as fs from 'fs';
import * as path from 'path';
import { FileContent } from '../../classes/FileContent/FileContent';
import { PathInfo } from '../../classes/PathInfo/PathInfo';

export function generateWorkspacePath(path: string) {
    return new PathInfo(`/home/document/${path}`);
}

export function getContentWorkspaceFile(filePath: string, ) {
    return fs.readFileSync( path.join(__dirname, 'base', filePath)).toString();
}

export function getExpectContent(filePath: string, ) {
    return fs.readFileSync( path.join(__dirname, 'expect', filePath)).toString();
}

export function getWorkspaceFile(filePath: string) {
    return new FileContent(generateWorkspacePath(filePath), getContentWorkspaceFile(filePath));
}

export const workspaceFile = [
    getWorkspaceFile('classes/footerClass.js'),
    getWorkspaceFile('classes/headerClass.js'),
    getWorkspaceFile('site/footerComponent/footer.html'),
    getWorkspaceFile('site/footerComponent/footer.scss'),
    getWorkspaceFile('site/headerComponent/header.html'),
    getWorkspaceFile('site/headerComponent/header.scss'),
];
