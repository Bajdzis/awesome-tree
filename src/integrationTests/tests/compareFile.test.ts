import { generateWorkspacePath, workspaceFile } from '../files/files';
import { FileContentCreator, FileContent } from '../../index';
import { AnalyzeFiles } from '../../classes/AnalyzeFiles/AnalyzeFiles';

describe('compareFile', () => {

    it('should generate info about files difference', () => {
        const generateFile = generateWorkspacePath('similar/xyzFile.txt');

        const similarFiles = workspaceFile.filter(file => generateFile.isSimilar(file.getPathInfo()));

        expect(similarFiles).toHaveLength(4);

        const comparer = new AnalyzeFiles();

        similarFiles.forEach(file => {
            const contentCreator = new FileContentCreator(generateFile, file);
            const newFileContent = new FileContent(generateFile, contentCreator.createContent());

            comparer.addFile(newFileContent);
        });

        expect(comparer.analyze()).toMatchSnapshot();
    });

});
