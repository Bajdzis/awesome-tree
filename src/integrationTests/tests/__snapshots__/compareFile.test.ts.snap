// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compareFile should generate info about files difference 1`] = `
FileContentNode {
  "body": Array [
    FileContentNodeAnalyze {
      "body": Array [
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "xyz_file
",
        },
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "",
        },
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "xyz_file
",
        },
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "some additional line
",
        },
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "",
        },
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "ggg_file
",
        },
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "",
        },
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "xyz_file
",
        },
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "some additional line
",
        },
        FileContentNodeAnalyze {
          "body": Array [],
          "end": "",
          "paths": Array [
            PathInfo {
              "parts": Array [
                WordsInfo {
                  "parts": Array [
                    "home",
                  ],
                  "words": "home",
                },
                WordsInfo {
                  "parts": Array [
                    "document",
                  ],
                  "words": "document",
                },
                WordsInfo {
                  "parts": Array [
                    "similar",
                  ],
                  "words": "similar",
                },
                WordsInfo {
                  "parts": Array [
                    "xyz",
                    "file",
                  ],
                  "words": "xyzFile",
                },
                WordsInfo {
                  "parts": Array [
                    "txt",
                  ],
                  "words": "txt",
                },
              ],
              "path": "/home/document/similar/xyzFile.txt",
              "type": "file",
            },
          ],
          "start": "",
        },
      ],
      "end": "",
      "paths": Array [
        PathInfo {
          "parts": Array [
            WordsInfo {
              "parts": Array [
                "home",
              ],
              "words": "home",
            },
            WordsInfo {
              "parts": Array [
                "document",
              ],
              "words": "document",
            },
            WordsInfo {
              "parts": Array [
                "similar",
              ],
              "words": "similar",
            },
            WordsInfo {
              "parts": Array [
                "xyz",
                "file",
              ],
              "words": "xyzFile",
            },
            WordsInfo {
              "parts": Array [
                "txt",
              ],
              "words": "txt",
            },
          ],
          "path": "/home/document/similar/xyzFile.txt",
          "type": "file",
        },
      ],
      "start": "",
    },
  ],
  "end": "",
  "start": "",
}
`;
